core/src/test/scala/kafka/server/metadata/MetadataPartitionsTest.scala

```diff
@@ -185,6 +184,134 @@ class MetadataPartitionsTest {
+    assertEquals(Set.empty, localRemoved(builder))
+    assertEquals(Set.empty, localChanged(builder))
+
+    val metadata = builder.build()
+    assertEquals(Set.empty, metadata.allTopicNames())
+    assertEquals(None, metadata.topicIdToName(topicId))
+    assertEquals(Set.empty, metadata.topicPartitions(topic).toSet)
+  }
+
+  @Test
+  def testTopicRemoval(): Unit = {
+    val brokerId = 0
+    val topic = "foo"
+    val numPartitions = 3
+    val topicPartitions = (0 until numPartitions).map(new TopicPartition(topic, _)).toSet
+    val createBuilder = new MetadataPartitionsBuilder(brokerId, emptyPartitions)
+    val topicId = createTopic(topic, numPartitions, createBuilder)
+    val localTopicPartitions = localChanged(createBuilder)
+    val createMetadata = createBuilder.build()
+
+    assertTrue(localTopicPartitions.subsetOf(topicPartitions))
+    assertTrue(localTopicPartitions.nonEmpty)
+    assertNotEquals(topicPartitions, localTopicPartitions)
+
+    val deleteBuilder = new MetadataPartitionsBuilder(brokerId = 0, createMetadata)
+    deleteBuilder.removeTopicById(topicId)
+    assertEquals(None, deleteBuilder.topicIdToName(topicId))
+    assertEquals(Set.empty, filterPartitions(deleteBuilder, topicPartitions))
+    assertEquals(localTopicPartitions, localRemoved(deleteBuilder))
+    assertEquals(Set.empty, localChanged(deleteBuilder))
+
+    val deleteMetadata = deleteBuilder.build()
+    assertEquals(Set.empty, deleteMetadata.allTopicNames())
+    assertEquals(None, deleteMetadata.topicIdToName(topicId))
+    assertEquals(Set.empty, deleteMetadata.topicPartitions(topic).toSet)
+  }
+
+  @Test
+  def testTopicDeleteAndRecreate(): Unit = {
+    val topic = "foo"
+    val numPartitions = 3
+    val initialBuilder = new MetadataPartitionsBuilder(0, emptyPartitions)
+    val initialTopicId = createTopic(topic, numPartitions, initialBuilder)
+    val initialLocalTopicPartitions = initialBuilder.localChanged().map(_.toTopicPartition).toSet
+    val initialMetadata = initialBuilder.build()
+
+    val recreateBuilder = new MetadataPartitionsBuilder(brokerId = 0, initialMetadata)
+    recreateBuilder.removeTopicById(initialTopicId)
+    assertEquals(initialLocalTopicPartitions, localRemoved(recreateBuilder))
+
+    val recreatedNumPartitions = 10
+    val recreatedTopicId = createTopic(topic, recreatedNumPartitions, recreateBuilder)
+    val recreatedTopicPartitions = (0 until recreatedNumPartitions).map(new TopicPartition(topic, _)).toSet
+    val recreatedLocalTopicPartitions = localChanged(recreateBuilder)
+
+    assertTrue(recreatedLocalTopicPartitions.nonEmpty)
+    assertNotEquals(recreatedLocalTopicPartitions, recreatedTopicPartitions)
+    assertTrue(recreatedLocalTopicPartitions.subsetOf(recreatedTopicPartitions))
+    assertFalse(recreatedLocalTopicPartitions.subsetOf(initialLocalTopicPartitions))
+    assertEquals(Some(topic), recreateBuilder.topicIdToName(recreatedTopicId))
```

Comment:
Could you check whether the id related to "topic" is changed to "recreatedTopicId"?